File: src/main/java/com/saomc/saoui/themes/elements/Hud.java
package com.saomc.saoui.themes.elements;

import javax.xml.bind.annotation.XmlRootElement;
import java.util.HashMap;
import java.util.Map;

/**
 * Part of saoui by Bluexin.
 *
 * @author Bluexin
 */
@XmlRootElement(namespace = "http://www.bluexin.be/com/saomc/saoui/hud-schema")
public class Hud {

    private Map<HudPartType, HudPart> parts = new HashMap<>();

    public Hud(Map<HudPartType, HudPart> parts) {
        this.parts = parts;
    }

    @SuppressWarnings("unused")
    protected Hud() {
    }

    public HudPart get(HudPartType key) {
        return parts.get(key);
    }

    public void setup() {
        this.parts.values().forEach(HudPart::setup);
    }
}

File: src/main/java/com/saomc/saoui/themes/elements/Hud.kt
package com.saomc.saoui.themes.elements

import com.saomc.saoui.api.themes.IHudDrawContext
import java.util.*
import javax.xml.bind.annotation.XmlRootElement

/**
 * Part of saoui by Bluexin.

 * @author Bluexin
 */
@XmlRootElement(namespace = "http://www.bluexin.be/com/saomc/saoui/hud-schema")
open class Hud protected constructor(override val name: String = "MenuDefs") : ElementParent {
    override fun getX(ctx: IHudDrawContext) = 0.0

    override fun getY(ctx: IHudDrawContext) = 0.0

    override fun getZ(ctx: IHudDrawContext) = 0.0

    private val parts = HashMap<HudPartType, ElementGroup>()

    operator fun get(key: HudPartType) = parts[key]

    fun setup() = this.parts.values.forEach { it.setup(this) }

    fun draw(key: HudPartType, ctx: IHudDrawContext) = this[key]?.draw(ctx)
}
